DIPs/DIP1xxx-rn.md

```diff
@@ -0,0 +1,801 @@
+phases (even if the threads don't actually touch each other's data) at the cost
+of performance. Python implements a global interpreter lock and it was proven
+to cause unscalable high contention; there are ongoing discussions of removing it
+from the Python implementation [5].
+
+### Introducing the copy constructor
+
+As stated above, the fundamental problem with the postblit is the automatic blitting
+of fields which makes it impossible to type check and cannot be synchronized without
+additional costs.
+
+As an alternative, this DIP proposes the implementation of a copy constructor. The benefits
+of this approach are the following:
+
+* it is a known concept. C++ has it [3];
+* it can be typechecked as a normal constructor (since no blitting is done, the data is
+initialized as if it were in a normal constructor); this means that `const`/`immutable`/`shared`
+copy constructors will be type checked exactly as their analogous constructors
+* offers encapsulation
+
+The downside of this solution is that the user must do all the field copying by hand
+and every time a field is added to a struct, the copy constructor must be modified.
+However, this can be easily bypassed by D's introspection mechanisms. For example,
+this simple code may be used as a language idiom or library function:
+
+```D
+foreach (i, ref field; src.tupleof)
+    this.tupleof[i] = field;
+```
+
+As shown above, the single benefit of the postblit can be easily substituted with a
+few lines of code. On the other hand, to fix the limitations of the postblit it is
+required that more complexity is added for little to no benefit. In these circumstances,
+for the sake of uniformity and consistency, replacing the postblit constructor with a
+copy constructor is the reasonable thing to do.
+
+## Description
+
+This section discusses all the technical aspects regarding the semantics
+of the copy constructor.
+
+### Syntax
+
+A declaration is a copy constructor declaration if it is a constructor declaration that takes only one
+parameter by reference that is of the same type as `typeof(this)`. Declaring a copy constructor as
+mentioned has the advantage that the parser does not get modified at all, thus leaving the language grammar
+unchanged.
+
+```d
+import std.stdio;
+
+struct A
+{
+    this(ref A rhs) { writeln("x"); }     // copy constructor
+}
+
+void main()
+{
+    A a;
+    A b = a;        // calls copy constructor implicitly
+    A c = A(b);     // calls constructor explicitly
+}
+```
+
+The copy constructor may also be called explicitly since it is also a constructor.
+
+The argument to the copy constructor is passed by reference in order to avoid infinite recursion (passing by
+value would require a copy of the `struct` which would be made by calling the copy constructor, thus leading
+to an infinite chain of calls).
+
+The type qualifiers may be applied to the parameter of the copy constructor, but also to the function itself
```

Comment:
The type qualifiers -> Type qualifiers

constructor, but also to -> constructor, and to

itself in -> itself, in